### PR TITLE
Now user can pass a custom time format

### PR DESCRIPTION
**Pull request by Punyaslok**
Now user can pass a custom time format according to which he wants his time to be parsed in case there isn't a match with existing time templates.

In cases where we get a "not a valid time string" error, we can explicitly give the time format.

`database.add_from_dir(sampledata_dir, ignore_already_added=True)` gives `ValueError: 01/06/2012 is not a valid time string!`

Now `database.add_from_dir(sampledata_dir, ignore_already_added=True, time_string_parse_format='%d/%m/%Y')` can be used to parse this time correctly.

### CODE REVIEW

**Comment by ehsteve** [comment]
This functionality is the same as that already provided by datetime.strptime. The original intent of parse_time is that it would try to figure the time format for you.

**Comment by Punyaslok** [comment]
@ehsteve parse_time() only supports certain formats given by [this](https://github.com/sunpy/sunpy/blob/master/sunpy/time/time.py#L27).

Dates like '01/07/2014' are not there as they are ambiguous. So this extra argument will parse this string successfully if the user passes '%d / %m / %Y' as its value. Here the user explicitly states that day comes before month.

Currently the behavior is ` ValueError: '01/07/2014' is not a valid time string!`

**Comment by Cadair** [comment]
Specifically this is motivated by GOES data.

**Comment by ehsteve** [comment]
True but datetime already provides exactly that functionality.

**Comment by Cadair** [comment]
@ehsteve, sure, but the point is not that datetime provides it (and we are just calling that functionality underneath). The point is that the user has no control over what they are passing to `parse_time` (it is being called by the database code many layers deep with the information from the header) all we are providing is a way for the user to manually say this file has `DATE_OBS` that has this format.

**Comment by ehsteve** [comment]
@Cadair I would suggest that we solve this problem in another way rather then through expanding the capabilities of parse_time. Perhaps there should be a database of time string formats for each instrument that we refer to and we just use datetime? This is the path of ssw, adding capabilities to functions, making them more confusing to use and generally leading to a giant mess.

**Comment by ehsteve** [comment]
@Punyaslok sorry to be so down on your PR. I'm glad it is forcing this conversation which is a very useful one.

**Comment by Cadair** [comment]
@Punyaslok we talked about this at the meeting. @ehsteve agreed that we should go ahead with this as a stop-gap. However, we should not make the `parse_time` keyword argument hidden (starts with `_` and lives in `kwargs`) so that we are not adding it to the public API.

**Comment by ayshih** [comment]
@Cadair Typo of "...we should ~not~ make..."?

**Comment by Punyaslok** [comment]
@ayshih Wait, so hidden or not hidden ?

**Comment by Cadair** [comment]
@Punyaslok it should be hidden. i.e. not part of the public API, my bad.

**Comment by Cadair** [comment]
This also needs a test for `parse_time` and a changelog entry (for the database api change). tests for database can be added once timeseries is merged and the culprit data is in master.

File: sunpy/database/tables.py [hunk 1]
```diff
@@ -432,7 +432,8 @@ def entries_from_query_result(qr, default_waveunit=None):
         yield DatabaseEntry._from_query_result_block(block, default_waveunit)
 
 
-def entries_from_file(file, default_waveunit=None):
+def entries_from_file(file, default_waveunit=None,
+        time_string_parse_format=None):
```
Comment (Cadair) [inline sunpy/database/tables.py]: indentation mis-match

File: sunpy/database/tables.py [hunk 2]
```diff
@@ -526,14 +527,14 @@ def entries_from_file(file, default_waveunit=None):
             # NOTE: the key DATE-END or DATE_END is not part of the official
             # FITS standard, but many FITS files use it in their header
             elif key in ('DATE-END', 'DATE_END'):
-                entry.observation_time_end = parse_time(value)
+                entry.observation_time_end = parse_time(value, _time_string_parse_format=time_string_parse_format)
```
Comment (Cadair) [inline sunpy/database/tables.py]: line length?

File: sunpy/database/tables.py [hunk 3]
```diff
@@ -526,14 +527,14 @@ def entries_from_file(file, default_waveunit=None):
...
             elif key in ('DATE-OBS', 'DATE_OBS'):
-                entry.observation_time_start = parse_time(value)
+                entry.observation_time_start = parse_time(value, _time_string_parse_format=time_string_parse_format)
```
Comment (Cadair) [inline sunpy/database/tables.py]: line length?

File: sunpy/database/tables.py [hunk 4]
```diff
@@ -526,14 +527,14 @@ def entries_from_file(file, default_waveunit=None):
...
         yield entry
 
 
 def entries_from_dir(fitsdir, recursive=False, pattern='*',
-        default_waveunit=None):
+        default_waveunit=None, time_string_parse_format=None):
```
Comment (Cadair) [inline sunpy/database/tables.py]: indentation mis-match.

**Comment by Cadair** [comment]
Nice!

File: sunpy/database/tables.py [hunk 5]
```diff
@@ -433,7 +434,8 @@ def entries_from_query_result(qr, default_waveunit=None):
         yield DatabaseEntry._from_query_result_block(block, default_waveunit)
 
 
-def entries_from_file(file, default_waveunit=None):
+def entries_from_file(file, default_waveunit=None,
+                      time_string_parse_format=None):
     """Use the headers of a FITS file to generate an iterator of
     :class:`sunpy.database.tables.DatabaseEntry` instances. Gathered
     information will be saved in the attribute `fits_header_entries`. If the
```
Comment (Cadair) [inline sunpy/database/tables.py]: The new kwarg needs documentation.

File: sunpy/database/tables.py [hunk 6]
```diff
@@ -528,14 +530,20 @@ def entries_from_file(file, default_waveunit=None):
...
+                        value,
+                        _time_string_parse_format=time_string_parse_format
+                        )
         yield entry
 
 
 def entries_from_dir(fitsdir, recursive=False, pattern='*',
-        default_waveunit=None):
+                     default_waveunit=None, time_string_parse_format=None):
     """Search the given directory for FITS files and use the corresponding FITS
```
Comment (Cadair) [inline sunpy/database/tables.py]: needs docs here as well.

**Comment by Cadair** [comment]
@ehsteve could you review this please? I know we talked about it, I would like to get it in as it's blocking the timeseries branch.

**Comment by Cadair** [comment]
This passed everything but the mac build that seems to be backed up on Travis, so I cancelled it.

Seeing how this PR has already been discussed, I am making the executive decision to merge with only my review as it is needed to get the timeseries PR #1846 to build and I wish to have that building for the meeting this evening.

**Pull request merged**